### PR TITLE
KAFKA-8620: fix NPE due to race condition during shutdown while rebalancing

### PR DESCRIPTION
**Pull request by abbccdda**
We have detected a race condition under system test failure. The problem was that the task manager internal active tasks should be guarded against state changes on the stream thread. Could definitely consider other fixes but this is currently the make-sense one.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by abbccdda** [comment]
@mjsax @guozhangwang @ableegoldman @cadonna @bbejeck @vvcephei Call for review

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 1]
```diff
@@ -190,14 +190,20 @@ State setState(final State newState) {
             oldState = state;
 
             if (state == State.PENDING_SHUTDOWN && newState != State.DEAD) {
+                log.debug("Invalid transition from PENDING_SHUTDOWN to {}: " +
+                              "only DEAD state is a valid next state", newState);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:194]: This might be miss leading for users who don't know the details. What about:
```
log.debug("Ignoring request to transit from PENDING_SHUTDOWN to {}", newState);
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 2]
```diff
@@ -190,14 +190,20 @@ State setState(final State newState) {
...
                 // when the state is already in PENDING_SHUTDOWN, all other transitions will be
                 // refused but we do not throw exception here
                 return null;
             } else if (state == State.DEAD) {
+                log.debug("Invalid transition from DEAD to {}: " +
+                              "no valid next state after DEAD", newState);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:200]: Similar to above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 3]
```diff
@@ -190,14 +190,20 @@ State setState(final State newState) {
...
                 // when the state is already in NOT_RUNNING, all its transitions
                 // will be refused but we do not throw exception here
                 return null;
             } else if (state == State.PARTITIONS_REVOKED && newState == State.PARTITIONS_REVOKED) {
+                log.debug("Invalid transition from PARTITIONS_REVOKED to PARTITIONS_REVOKED: " +
+                              "self transition is not allowed");
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:206]: Thinking about this, I am wondering if we should just change the FSM to allow this transition and simplify the code here? \cc @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 4]
```diff
@@ -268,9 +274,13 @@ public void onPartitionsAssigned(final Collection<TopicPartition> assignment) {
             final long start = time.milliseconds();
             try {
                 if (streamThread.setState(State.PARTITIONS_ASSIGNED) == null) {
-                    return;
-                }
-                if (streamThread.assignmentErrorCode.get() == StreamsPartitionAssignor.Error.NONE.code()) {
+                    log.error("State transition from {} to PARTITIONS_ASSIGNED failed. " +
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Why is this an ERROR? If we receive a SHUTDOWN signal, it's just an "eager exit" to not create tasks IMHO. We might want to log a DEBUG though.

Would also update the message:
```
log.debug("Skipping task creation in rebalance because we are already in shutdown phase.");
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 5]
```diff
@@ -268,9 +274,13 @@ public void onPartitionsAssigned(final Collection<TopicPartition> assignment) {
...
+                                  "Will skip the task initialization", streamThread.state());
+                } else if (streamThread.assignmentErrorCode.get() != StreamsPartitionAssignor.Error.NONE.code()) {
+                    log.error("Encountered assignment error during partition assignment: {}. " +
+                                  "Will skip the task initialization", streamThread.assignmentErrorCode);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: There are two different error codes:
 - `VERSION_PROBING`: for this case we continue and rejoin the group; also the assignment is empty
 - `INCOMPLETE_SOURCE_TOPIC_METADATA`: for this case we shut down, and this case is already handled above.

Hence, I think we don't need to log anything for this case? (Note that we log the version probing in StreamThread later and the metadata error is logged above already)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 6]
```diff
@@ -830,6 +839,15 @@ void runOnce() {
             throw new StreamsException(logPrefix + "Unexpected state " + state + " during normal iteration");
         }
 
+        // Shutdown hook could potentially be triggered and transit the thread state to PENDING_SHUTDOWN during #pollRequests().
+        // The task manager internal states could be uninitialized if the state transition happens during #onPartitionAssigned().
+        // Should only proceed when the thread is still running after #pollRequests(), because no external state mutation
+        // could affect the task manager state beyond this point within #runOnce().
+        if (!isRunning()) {
+            log.warn("State already transits to {}, skipping the run once call after poll request", state);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Seems like this should be DEBUG level?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 7]
```diff
@@ -1179,7 +1197,7 @@ private void completeShutdown(final boolean cleanRun) {
         // intentionally do not check the returned flag
         setState(State.PENDING_SHUTDOWN);
 
-        log.info("Shutting down");
+        log.info("Shutting down with cleanRun {}", cleanRun);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Not sure if we need this? If the shutdown is not clean, we logged an ERROR before in `StreamThread#run()`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 8]
```diff
@@ -695,6 +695,58 @@ public void shouldShutdownTaskManagerOnCloseWithoutStart() {
...
+        internalTopologyBuilder.addSource(null, "source", null, null, null, topic1);
+
+        final StreamThread streamThread = createStreamThread(clientId, new StreamsConfig(configProps(true)), true);
+
+        final Map<TaskId, Set<TopicPartition>> activeTasks = new HashMap<>();
+        final List<TopicPartition> assignedPartitions = new ArrayList<>();
+
+        // assign single partition
+        assignedPartitions.add(t1p1);
+        assignedPartitions.add(t1p2);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Why do we need two partitions?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 9]
```diff
@@ -695,6 +695,58 @@ public void shouldShutdownTaskManagerOnCloseWithoutStart() {
...
+        activeTasks.put(task2, Collections.singleton(t1p2));
+
+        streamThread.taskManager().setAssignmentMetadata(activeTasks, Collections.emptyMap());
+        final MockConsumer<byte[], byte[]> mockConsumer = (MockConsumer<byte[], byte[]>) streamThread.consumer;
+        final Map<TopicPartition, Long> beginOffsets = new HashMap<>();
+        beginOffsets.put(t1p1, 0L);
+        beginOffsets.put(t1p2, 0L);
+        mockConsumer.updateBeginningOffsets(beginOffsets);
+
+        final Thread callbackThread = new Thread(() -> {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Why do we need a `Thread` here?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 10]
```diff
@@ -695,6 +695,58 @@ public void shouldShutdownTaskManagerOnCloseWithoutStart() {
...
+        });
+
+        streamThread.setStateListener(
+            (t, newState, oldState) -> {
+                if (oldState == StreamThread.State.CREATED && newState == StreamThread.State.STARTING) {
+                    // Start triggering consumer callbacks to make sure we proceed to PARTITIONS_ASSIGNED stage.
+                    callbackThread.start();
+                } else if (oldState == StreamThread.State.PARTITIONS_REVOKED && newState == StreamThread.State.PARTITIONS_ASSIGNED) {
+                    // Immediately trigger shutdown call to switch the state.
+                    streamThread.shutdown();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:727]: To test the race condition, should we call this when `newState == PARTITIONS_REVOKED`, ie, one state transition earlier?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 11]
```diff
@@ -695,6 +695,58 @@ public void shouldShutdownTaskManagerOnCloseWithoutStart() {
...
+                }
+            });
+
+        streamThread.run();
+
+        LogCaptureAppender.unregister(appender);
+        final List<String> strings = appender.getMessages();
+        assertTrue(strings.contains(
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Not sure if we need this? If the test is correct, wouldn't `streamThread.run();` not thrown an NPE without the fix?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:194]: Would be better to add an explanation for ignorance?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 4 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Sounds good to me.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 6 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Sounds good

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java (hunk 9 above)
Comment (abbccdda) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: We need a separate thread to trigger callbacks, otherwise the state won't proceed. Maybe there is a better way to do that?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java (hunk 10 above)
Comment (abbccdda) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:727]: We couldn't reproduce the NPE in `partitionRevoked` because it's not expected to return records from `poll`. We need to trigger shutdown exactly during the call the `partitionAssigned` so that we are expecting records to return to attempt adding records to task manager. 
However current solution is not stable either, because it's not a correct way to trigger the condition. The natural way should leverage mock consumer to trigger assignment calls. Maybe we could talk in person and see how to reproduce it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:194]: Sure. Why not. I just would avoid the term "invalid" because it might confuse users (they may thing something bad happens, but it's expected and not bad).

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java (hunk 9 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Can you elaborate? What do you mean by

> otherwise the state won't proceed

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 12]
```diff
@@ -268,9 +274,13 @@ public void onPartitionsAssigned(final Collection<TopicPartition> assignment) {
             final long start = time.milliseconds();
             try {
                 if (streamThread.setState(State.PARTITIONS_ASSIGNED) == null) {
-                    return;
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:271]: Why did you remove the `return`? Without the `return`, the code will reach the `finally` block and log that the assignment took place.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 13]
```diff
@@ -268,9 +274,13 @@ public void onPartitionsAssigned(final Collection<TopicPartition> assignment) {
...
-                }
-                if (streamThread.assignmentErrorCode.get() == StreamsPartitionAssignor.Error.NONE.code()) {
+                    log.error("State transition from {} to PARTITIONS_ASSIGNED failed. " +
+                                  "Will skip the task initialization", streamThread.state());
+                } else if (streamThread.assignmentErrorCode.get() != StreamsPartitionAssignor.Error.NONE.code()) {
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:281]: Isn't this a rather brittle fix? How does this code guarantee that after `streamThread.assignmentErrorCode.get() != StreamsPartitionAssignor.Error.NONE.code()` is `false`, the result of `streamThread.assignmentErrorCode.get()` does not change again before the tasks are created?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 7 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I would not use variable names in log messages.

**Review by cadonna — COMMENTED** [review]
@abbccdda, Thank you for the PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:194]: Sounds good

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 12 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:271]: I'm confused, we will reach `finally` block no matter we have `return` or or not right?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 5 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I guess it's still useful to have a debug log here as it helps us understand why we skip the assignment.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 7 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I could avoid using variable name in the log, but still feel needed to log the flag here, as we can't control future changes to set cleanRun flag.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 12 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:271]: Oh sorry, I misunderstood the code.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 13 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:281]: in `PartitionsAssigned` we should have already passed the assignment error placement, so I won't expect another flip for its value during this round rebalance.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 14]
```diff
@@ -268,17 +274,19 @@ public void onPartitionsAssigned(final Collection<TopicPartition> assignment) {
             final long start = time.milliseconds();
             try {
                 if (streamThread.setState(State.PARTITIONS_ASSIGNED) == null) {
-                    return;
-                }
-                if (streamThread.assignmentErrorCode.get() == StreamsPartitionAssignor.Error.NONE.code()) {
+                    log.debug("Skipping task creation in rebalance because we are already in {} phase.",
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]:
```suggestion
                    log.debug("Skipping task creation in rebalance because we are already in {} state.",
```

(minor suggestion)

**Review by vvcephei — APPROVED** [review]
LGTM. Just one minor suggestion.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 13 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:281]: Makes sense

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 5 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Well. Its double logging... Not a big fan of that. But if you insist to keep if, also fine.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 15]
```diff
@@ -1179,7 +1195,7 @@ private void completeShutdown(final boolean cleanRun) {
         // intentionally do not check the returned flag
         setState(State.PENDING_SHUTDOWN);
 
-        log.info("Shutting down");
+        log.info("Shutting down with clean flag to {}", cleanRun);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: nit: `clean flag` sound like an implementation detail, that may not be a good way to phrase it for an INFO log...

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 16]
```diff
@@ -1570,4 +1611,30 @@ private void addRecord(final MockConsumer<byte[], byte[]> mockConsumer,
...
+        }
+
+        @Override
+        public synchronized ConsumerRecords<K, V> poll(final Duration timeout) {
+            assertNotNull(streamThread);
+            onPoll();
+            return super.poll(timeout);
+        }
+
+        public abstract void onPoll();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Why do we need this method? Can we call `streamThread.shutdown();` directly in `poll(...)` ?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 17]
```diff
@@ -695,6 +697,45 @@ public void shouldShutdownTaskManagerOnCloseWithoutStart() {
         EasyMock.verify(taskManager);
     }
 
+    @Test
+    public void shouldNotAccessTaskManagerWhenPendingShutdownInRunOnce() {
+        final TaskManager taskManager = EasyMock.createNiceMock(TaskManager.class);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Why do we need to mock the taskManager?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 15 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: cool, will remove that

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java (hunk 16 above)
Comment (abbccdda) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Just want to have this mock consumer being reused later if someone needs to inject different behavior

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java (hunk 17 above)
Comment (abbccdda) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Just to simplify the setup, otherwise we have to get a bunch of parameters in order to create one.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 5 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Addressed

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 18]
```diff
@@ -695,6 +698,85 @@ public void shouldShutdownTaskManagerOnCloseWithoutStart() {
         EasyMock.verify(taskManager);
     }
 
+    @Test
+    public void shouldNotAccessTaskManagerWhenPendingShutdownInRunOnce() {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: I would give this test a more meaningful name, e.g. `shouldNotThrowWhenPendingShutdownInRunOnce`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 19]
```diff
@@ -695,6 +698,85 @@ public void shouldShutdownTaskManagerOnCloseWithoutStart() {
...
+        mockRunOnce(true);
+    }
+
+    @Test
+    public void shouldHaveNoErrorAccessTaskManagerInRunOnceNoShutdown() {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Same as above, e.g. `shouldNotThrowWithoutPendingShutdownInRunOnce`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 20]
```diff
@@ -268,17 +274,19 @@ public void onPartitionsAssigned(final Collection<TopicPartition> assignment) {
             final long start = time.milliseconds();
             try {
                 if (streamThread.setState(State.PARTITIONS_ASSIGNED) == null) {
-                    return;
-                }
-                if (streamThread.assignmentErrorCode.get() == StreamsPartitionAssignor.Error.NONE.code()) {
+                    log.debug("Skipping task creation in rebalance because we are already in {} state.",
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Shouldn't the indentation be as follows?
```
log.debug(
    "Skipping task creation in rebalance because we are already in {} state.",
    streamThread.state()
);
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 21]
```diff
@@ -268,17 +274,19 @@ public void onPartitionsAssigned(final Collection<TopicPartition> assignment) {
...
+                              streamThread.state());
+                } else if (streamThread.assignmentErrorCode.get() != StreamsPartitionAssignor.Error.NONE.code()) {
+                    log.debug("Encountered assignment error during partition assignment: {}. " +
+                                  "Will skip the task initialization", streamThread.assignmentErrorCode);
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Shouldn't the indentation be as follows?
```
log.debug(
    "Encountered assignment error during partition assignment: {}. Will skip the task initialization", 
    streamThread.assignmentErrorCode
);
```
The first parameter is one or two characters too long, though.

**Review by cadonna — COMMENTED** [review]
LGTM
@mjsax, What is so bad about using a mock for the task manager?

**Review by cadonna — APPROVED** [review]

**Comment by cadonna** [comment]
Retest this, please

**Comment by abbccdda** [comment]
@cadonna Thank you for the review, addressed!

**Comment by mjsax** [comment]
@cadonna The bug is, that if we don't create tasks on the `TaskManager` and when we later try to receive tasks from it, we hit an NPE. Hence, the `TaskManager` is on the test code path -- mocking it seems to make the test weak IMHO.

**Review by mjsax — APPROVED** [review]

**Comment by abbccdda** [comment]
Retest this please

**Comment by abbccdda** [comment]
Retest this please

**Comment by cadonna** [comment]
@mjsax I do not agree. The `TaskManager` is on the test code path, but it does not need to be tested. A unit test should test the class under test which is `StreamThread` in this case and avoid as much as possible testing other classes. With a `TaskManager` mock, you could easily verify if the calls to the `TaskManager` that would result in a `NullPointerException` are done or not. In addition, with a `TaskManager` mock the tests would also better document what the productive code is supposed to do because you have to specify it explicitly on the mock.

**Comment by mjsax** [comment]
@cadonna Maybe. Do you want that we switch back to a mock? If yes, please let @abbccdda know. If not, please let me know so the PR can be merged. Your call.

**Comment by abbccdda** [comment]
Retest this please

**Comment by abbccdda** [comment]
@cadonna @mjsax My take is that the purpose of mocking taskManager is to create a conditional behavior path: if you set an active task, you could retrieve it later; if you don't, you will get NPE, while the setup of real taskManger is too much compared with a mock. Actually in our case, the mock examples in other tests doesn't quite fit to implement this mentioned conditional logic, and the task manager setup is not bad either. So we should be fine just creating a new client in this case.

**Comment by ijuma** [comment]
Reopening to trigger Jenkins build.

**Comment by cadonna** [comment]
@abbccdda: What you describe is also a valid use case of a mock, but it only achieves the isolation of the class under test. It does not document well the expected behaviour of productive code. Furthermore, IMO conditional behaviour in a mock should be avoided if possible, because it introduces code that needs maintenance and could be wrong. A mock should be as simple as possible. Admittedly, in this case the code would probably not be that complex, but if it could be avoided, I would avoid it. 

@mjsax @abbccdda: I do not want to block this fix further. The test seems to be correct. My concern is about clean and maintainable code. Up to you to follow my reasoning or not.

**Comment by bbejeck** [comment]
retest this please

**Comment by bbejeck** [comment]
retest this please

**Comment by bbejeck** [comment]
retest this please

**Comment by abbccdda** [comment]
Checked log and the single test failure was in RebalanceSourceConnectorsIntegrationTest in Connect, not related

**Comment by bbejeck** [comment]
retest this please

**Comment by abbccdda** [comment]
retest this please

**Comment by abbccdda** [comment]
retest this please

**Comment by bbejeck** [comment]
failures unrelated 
retest this please

**Comment by abbccdda** [comment]
All flaky tests:
ExampleConnectIntegrationTest > testSourceConnector

**Comment by abbccdda** [comment]
retest this please

**Comment by abbccdda** [comment]
@mjsax @bbejeck I have witnessed 4 times 2/3 green builds, and all failures are flaky tests including this time: `ExampleConnectIntegrationTest.testSourceConnector`
I personally don't think there is any further meaning to trigger another jenkins build, but you will make the call.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:206]: I thought about this when tightening the FSM before but the unit tests reminds me of one thing: our current contract is that we only transit to PARTITIONS_REVOKED when calling onPartitionsRevoked, which is called only once at the beginning of the rebalance today, so keeping it strict is better just in case we have incorrect partial rebalance procedure.

With KIP-429 this may be violated so we need to revisit our FSM once Streams adopt cooperative protocols. cc @ableegoldman who's working on this.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 22]
```diff
@@ -268,17 +274,23 @@ public void onPartitionsAssigned(final Collection<TopicPartition> assignment) {
...
+                    log.debug("Creating tasks based on assignment.");
                     taskManager.createTasks(assignment);
                 }
             } catch (final Throwable t) {
                 log.error(
                     "Error caught during partition assignment, " +
-                        "will abort the current process and re-throw at the end of rebalance: {}",
-                    t
-                );
+                        "will abort the current process and re-throw at the end of rebalance", t);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:293]: Nice catch.

**Review by guozhangwang — APPROVED** [review]

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.3`, `2.2`, and `2.1` branches.
